parser-typechecker/src/Unison/Codebase/SqliteCodebase/Operations.hs

```diff
@@ -524,3 +534,62 @@ declExists = termExists
 before :: Branch.Hash -> Branch.Hash -> Transaction (Maybe Bool)
 before h1 h2 =
   Ops.before (Cv.causalHash1to2 h1) (Cv.causalHash1to2 h2)
+
+namesWithinPath ::
+  Maybe Path ->
+  Transaction ScopedNames
+namesWithinPath path = do
+  (termNames, typeNames) <- Ops.rootBranchNames
+  let allTerms :: [(Name, Referent.Referent)]
+      allTerms =
+        termNames <&> \(S.NamedRef {reversedSegments, ref = (ref, ct)}) ->
+          let v1ref = runIdentity $ Cv.referent2to1 (const . pure . Cv.constructorType2to1 . fromMaybe (error "Required constructor type for constructor but it was null") $ ct) ref
+           in (Name.fromReverseSegments (coerce reversedSegments), v1ref)
+  let allTypes :: [(Name, Reference.Reference)]
+      allTypes =
+        typeNames <&> \(S.NamedRef {reversedSegments, ref}) ->
+          (Name.fromReverseSegments (coerce reversedSegments), Cv.reference2to1 ref)
+  let rootTerms = Rel.fromListDomainAsc allTerms
+  let rootTypes = Rel.fromListDomainAsc allTypes
+  let absoluteRootNames = Names {terms = rootTerms, types = rootTypes}
+  let (relativeScopedNames, absoluteExternalNames) =
+        case path of
+          Just p@(_ Path.:> _) ->
+            let reversedPathSegments = reverse . Path.toList $ p
+                (relativeTerms, externalTerms) = foldMap (partitionByPathPrefix reversedPathSegments) allTerms
+                (relativeTypes, externalTypes) = foldMap (partitionByPathPrefix reversedPathSegments) allTypes
+             in ( Names {terms = Rel.fromListDomainAsc relativeTerms, types = Rel.fromListDomainAsc relativeTypes},
+                  Names {terms = Rel.fromListDomainAsc externalTerms, types = Rel.fromListDomainAsc externalTypes}
+                )
+          _ -> (absoluteRootNames, mempty)
+  pure $
+    ScopedNames
+      { absoluteExternalNames,
+        relativeScopedNames,
+        absoluteRootNames
+      }
+  where
+    partitionByPathPrefix :: [NameSegment] -> (Name, r) -> ([(Name, r)], [(Name, r)])
```

Comment:
Could you add a comment with a couple examples showing what this function does?

Reply:
Also, I think a `DList` would work better, since we are folding over all terms and types in the namespace. Can be in another PR, just addressed by a FIXME for now

Reply:
Good thought though! TIL about strict left-associative `foldMap'`; I knew about all the other strict folds but didn't know `foldMap` had one haha.